{% extends "base.html" %}
{% load staticfiles %}
{% block title %}TFBS - Home{% endblock %}
{% block head %}{{ block.super }}
    <script type="text/javascript">
      function tfAutocomplete() {
        $('#term').autocomplete({
          source: '/tfcompletions',
          minLength: 2
        });
      }
      function geneAutocomplete() {
        $('#term').autocomplete({
          source: '/genecompletions',
          minLength: 2
        });
      }
      $(document).ready(function() {
        $('#stype').change(function() {
          var stype = $(this).val();
          if (stype == 'tf') {
            tfAutocomplete();
            $('#searchform').attr('action', 'searchtf');
          } else {
            geneAutocomplete();
            $('#searchform').attr('action', 'searchgene');
          }
        });
        tfAutocomplete();
      });
    </script>{% endblock %}
    {% block content %}
    <p align='left'>This database of transcription factor (TF) to target gene interactions was constructed by combining: 1) human genome sequence, 2) 2,331 unique DNA recognition motifs for 690 TFs from human and mouse genomes (<a href='http://www.ncbi.nlm.nih.gov/pubmed/16381825'>Matys et al., 2006</a>; <a href='http://www.ncbi.nlm.nih.gov/pubmed/18842628'>Newburger and Bulyk, 2009</a>; <a href='http://www.ncbi.nlm.nih.gov/pubmed/23332764'>Jolma et al., 2013</a>; <a href='http://www.ncbi.nlm.nih.gov/pubmed/24194598'>Mathelier et al., 2014</a>), and 3) occupancy of 8.4 million genomic sites through ENCODE DNase I footprints across 41 diverse cell and tissue types (<a href='http://www.ncbi.nlm.nih.gov/pubmed/22955618'>Neph et al., 2012</a>). In total 17,415,125 significant motif instances (FIMO p-value = 1 x 10<sup>-5</sup>) were discovered for the 2,331 TF DNA recognition motifs using an optimal promoter size of &#177;5Kbp from the transcriptional start site. A quarter of these motif instances (3,505,491 motif instances) that overlapped by at least 1bp with a digital genomic footprint (DGF) were used to construct a mechanistic regulatory network of 690 TFs regulating 18,153 genes.</p>
    <center>
        <table>
            <tr>
                <td>Construction of  genome-wide TF to target gene database using by integrating genomic sequence, TF DNA recognition motifs, and DGF footprints. <b>A.</b> Promoter regions of genes (±5Kbp of the transcriptional start site (TSS)) were searched for DNA sequences that significantly matched a TF DNA recognition motif. As an example we use the TF USF1 which has three DNA sequences (motif instances) that significantly match the USF1 DNA recognition motif. The USF1 protein binds to only one of these motif instances, but from sequence alone it is not possible to discriminate between functional and non-functional motif instances. The DNAse I hypersensitivity hotspots from DNAse I-seq provides information about which regions of a promoter have open chromatin. However, presence of a DNAse hypersensitivity hotspot does not directly imply a motif instance will be bound by its cognate TF. Digtial Genomic Footprinting (DGF) provides a means to empirically determine the genome-wide occupancy of DNA by TFs and other factors at nucleotide resolution. Using the additional evidence provided by DGF footprints that something was bound to the same location as a motif instance allows us to exclude a large fraction of motif instances that aren’t bound by a TF. <b>B.</b> ChIP-seq with antibodies specific to a particular TF provides a means to empirically determine the genome-wide binding locations for a specific TF. We compare our predicted TF to target gene interactions to specific binding patterns provided by ChIP-seq studies. <b>C.</b> Filtering predicted transcription factor binding sites through DNAse I hypersensitivity footprints significantly increases sensitivity and specificity for predicting TF from ENCODE ChIP-seq studies. Each line is a ROC curve for comparison to the corresponding the ENCODE ChIP-seq studies. AUC = area under the curve. <b>D.</b> Comparisons of ROC AUCs from increasing upstream promoter lengths were made relative to the core promoter size of ±500bp. A promoter length exceeding the red line indicates a significant reduction in ROC AUC (p-value &lt; 0.05). <b>E.</b> Comparisons of ROC AUCs from increasing downstream promoter lengths were made relative to the promoter size of -5Kbp and +500bp. A promoter length exceeding the red line indicates a significant reduction in ROC AUC (p-value &lt; 0.05).</td>
            </tr>
            <tr>
                <td><img src='{% static "images/constructionValidation.gif" %}'></td>
            </tr>
        </table>
    </center>
    
    <center><p><b>Database statistics</b></br>
    <u>Number of genes:</u> {{num_genes}}</br>
    <u>Number of motifs:</u> {{num_motifs}}</br>
    <u>Number of TFBS:</u> {{num_tfbs}}</p>
    <div id="results"></div>
    </center>
    {% endblock %}
